compiler/AST/expr.cpp

```diff
@@ -5839,383 +5838,379 @@ GenRet CallExpr::codegenPrimMove() {
+
+    codegenAssign(to_ptr, get(2));
+
+  } else if (get(1)->typeInfo()->symbol->hasFlag(FLAG_REF)        == true  &&
+             get(2)->typeInfo()->symbol->hasFlag(FLAG_WIDE_REF)   == true)  {
+    if (get(1)->getValType() != get(2)->getValType()) {
+      GenRet narrowRef   = codegenRaddr(get(2));
+      GenRet wideThing   = codegenDeref(narrowRef);
+      GenRet narrowThing = codegenWideThingField(wideThing, WIDE_GEP_ADDR);
+
+      codegenAssign(get(1), codegenAddrOf(narrowThing));
+    } else {
+      codegenAssign(get(1), codegenRaddr(get(2)));
+    }
+
+  } else if (get(1)->typeInfo()->symbol->hasFlag(FLAG_WIDE_CLASS) == false &&
+             get(1)->typeInfo()->symbol->hasFlag(FLAG_REF)        == false &&
+             get(2)->typeInfo()->symbol->hasFlag(FLAG_WIDE_CLASS) == true)  {
+    codegenAssign(get(1), codegenRaddr(get(2)));
+
+  } else if (get(1)->typeInfo()->symbol->hasFlag(FLAG_REF)        == true  &&
+             get(2)->typeInfo()->symbol->hasFlag(FLAG_REF)        == false) {
+    codegenAssign(codegenDeref(get(1)), get(2));
+
   } else {
-    if (CallExpr* call = toCallExpr(get(2))) {
-      if (call->primitive) {
-        // Assume that the RHS is a primitive we wish to handle specially
-        // until proven otherwise.
-        // If the RHS is not handled specially, we fall through and handle it
-        // generally.
-        bool handled = true;
-
-        switch (call->primitive->tag) {
-        case PRIM_GET_REAL:
-        case PRIM_GET_IMAG: {
-          bool isReal = call->primitive->tag == PRIM_GET_REAL;
-
-          if (call->get(1)->typeInfo()->symbol->hasFlag(FLAG_WIDE_REF)) {
-            // move(wide_real, prim_get_real(wide_complex));
-            // turns into: wide_real.locale = wide_complex.locale;
-            //             wide_real.addr = prim_get_real(wide_complex.addr);
-            Type*  cplxType = call->get(1)->typeInfo()->getRefType();
-
-            GenRet t1       = createTempVar(cplxType);
-            codegenAssign(t1, codegenRaddr(call->get(1)));
-
-            GenRet t2       = createTempVar(get(1)->typeInfo()->getRefType());
-            codegenAssign(t2, codegen_prim_get_real(t1, cplxType, isReal));
-
-            GenRet to_ptr   = get(1);
-            GenRet from     = codegenWideAddr(codegenRlocale(call->get(1)),
-                                              codegenDeref(t2));
-
-            if (gGenInfo->cfile) {
-              std::string stmt = codegenValue(to_ptr).c + " = ";
-
-              stmt += from.c;
-              stmt += ";\n";
-
-              gGenInfo->cStatements.push_back(stmt);
-            } else {
-#ifdef HAVE_LLVM
-              codegenStoreLLVM(from, to_ptr);
-#endif
-            }
+    codegenAssign(get(1), get(2));
+  }
 
-          } else {
-            codegenAssign(get(1), codegen_prim_get_real(call->get(1), call->get(1)->typeInfo(), isReal));
-          }
+  return ret;
+}
 
-          break;
-        }
+bool CallExpr::codegenPrimMoveSpecial() {
```

Comment:
I would appreciate a comment saying something about what makes these PRIM_MOVE's "special".

Here is a 1st stab at what I might write:

While generally speaking PRIM_MOVE works with a left-hand-side variable and a right-hand-side CallExpr, many right-hand-side CallExprs receive special treatment so that they can affect the movement of data from the right-hand-side to the left-hand-side of the move. The most common adjustment here is to implement the PRIM_MOVE itself with a GET instead of the usual local shallow copy.

Of course you are free to decline to do such a comment in this PR.